gossip/src/crds_value.rs

```diff
@@ -494,27 +497,53 @@ impl Sanitize for NodeInstance {
 pub struct RestartLastVotedForkSlots {
     pub from: Pubkey,
     pub wallclock: u64,
-    pub slots: Vec<CompressedSlots>,
+    pub compressed_slots: CompressedSlots,
     pub last_voted_hash: Hash,
     pub shred_version: u16,
 }
 
 impl Sanitize for RestartLastVotedForkSlots {
     fn sanitize(&self) -> std::result::Result<(), SanitizeError> {
-        if self.slots.is_empty() {
+        if self.compressed_slots.num_slots() == 0 {
             return Err(SanitizeError::InvalidValue);
         }
-        self.slots.sanitize()?;
+        self.compressed_slots.sanitize()?;
         self.last_voted_hash.sanitize()
     }
 }
 
 impl RestartLastVotedForkSlots {
-    pub fn new(from: Pubkey, now: u64, last_voted_hash: Hash, shred_version: u16) -> Self {
+    pub fn new(
+        from: Pubkey,
+        now: u64,
+        last_voted_fork: &mut [Slot],
+        last_voted_hash: Hash,
+        shred_version: u16,
+    ) -> Self {
+        last_voted_fork.sort();
+        let mut compressed_slots = CompressedSlots::new(1);
+        let fork_length = last_voted_fork.len();
+        let last_voted_slot = last_voted_fork.last();
+        for slice_length in
+            max(1, min(fork_length, MAX_RESTART_LAST_VOTED_FORK_SLOTS_SPACE))..fork_length + 1
+        {
+            let mut new_slots = CompressedSlots::new(slice_length);
+            let new_slice = &last_voted_fork[fork_length.saturating_sub(slice_length)..];
+            new_slots.add(new_slice);
+            let _ = new_slots.deflate();
+            if last_voted_slot.unwrap() - new_slots.first_slot() < MAX_SLOTS_PER_ENTRY as u64
+                && serialized_size(&new_slots).unwrap()
+                    < MAX_RESTART_LAST_VOTED_FORK_SLOTS_SPACE as u64
+            {
+                compressed_slots = new_slots;
+            } else {
+                break;
+            }
+        }
```

Comment:
This is now pretty confusing and hacky. Can we do something better in here?

I think part of the complexity is because of using `CompressedSlots` here, but we need a much simpler construct here because unlike `EpochSlots` this is not updating/appending to old values.

For example, just encoding the last slot, and a bit-vector for representing the parents.

Reply:
Since we only have ~800 bytes available, a straight bit-vector can only encode 7.2k slots, so I put flate2 compression in there.

Reply:
Are you sure that flate2 compression don't make it worse?
I mean the bits representing parent slots can be totally random and not have any patterns, in which case compression could make it worse.
Also there is this concern with using compression: https://github.com/solana-labs/solana/pull/14362#discussion_r550805153

Reply:
That is valid concern. The problem of not using compression is that we can only have 7.2k slots going back in time which is roughly 48 minutes, this is really too short because currently we rarely resolve an outage within 48 minutes. (Even if your local vote is chaining to a parent generated 2 hours ago, your bitmap in the past 48 minutes will all be 0's so it's zero information.)

About whether compression will make it worse, I tested a few case, it sounds like compression can compress large sequence of 0's or 1's really well. If you think about what's happening in an outage (@carllin since he expressed similar concerns):
1. outages where blocks arrive very slowly (outage 20230225, 20220430): most validators can't really vote on the blocks generated by others, so it's voting on a local fork with blocks generated by itself
2. outages where new blocks can't chain off each other, they all keep building from the parent (outage 20220930): same as above, voting on a local fork with blocks generated by itself
3. outages where >1/3 of the validators don't agree any more (outage 20220601): in this case we normally have several big forks, so you should see >1/3 of 1's in the forks we care about, that should be many consecutive 1's

So I think in the outage scenarios we have seen, compression will generally help. If we are concerned about compression here, we can implement our own compression in a separate PR if needed, or we can consider more than one RestartLastVotedForkSlots packets.

Reply:
if you expect a lot of consecutive `1`'s in the parent chain, then maybe we can use a run-length encoding for `1`s 
https://en.wikipedia.org/wiki/Run-length_encoding
The lengths can also be encoded with varint encoding:
https://github.com/solana-labs/solana/blob/54b796f5a/sdk/program/src/serde_varint.rs

I don't know if it is going to be better or not, and the worst case still adds overhead compared to not encoded raw data. 
do you want to do some experiments on how does above compare against flate2 compression?

Reply:
I think during outages there will be lots of consecutive 0's or consecutive 1's in the bitmap, consecutive 0's happen when blocks arrive very slowly, consecutive 1's happen when >1/3 of the validators diverge. In both cases run-length encoding can probably compress the bitmap tight enough to fit 16k slots. I'll have a try at implementing a very simple run-length encoding.

Reply:
I replaced flate2 compression with a simple run-length encoding implementation. Under the best case (tons of consecutive 0's or 1's) the compression rate is 127:1. Under the worst case (very random) it performs worse than raw bits. The performance is comparable to flate2, but it's so much easier to fit everything into given byte array here. The previous flate2 implementation has a bug because at end of compression it needs to write some headers into the output buffer, and the header length is very hard to predict.

Reply:
And RestartLastVotedForkSlots will select raw bits if that performs better.